src/libcore/str/pattern.rs

```diff
@@ -824,27 +843,51 @@ impl TwoWaySearcher {
         // "Algorithm CP2", which is optimized for when the period of the needle
         // is large.
         if &needle[..crit_pos] == &needle[period.. period + crit_pos] {
-            // short period case
+            // short period case -- the period is exact
+            let byteset = needle[..period].iter()
+                                .fold(0, |a, &b| (1 << (b & 0x3f)) | a);
```

Comment:
Maybe there should be some method (probably next to `byteset_contains`) for constructing the byteset?